Remove explicit flushes from HTTP2 encoders, decoders & flow-controllers

louiscryan opened the pull request:
@Scottmitch 

Mostly fixes https://github.com/netty/netty/issues/3670

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2RemoteFlowController.java:
```diff
@@ -27,8 +27,8 @@
-     * Manually flushing the {@link ChannelHandlerContext} is not required, since the flow
-     * controller will flush as appropriate.
+     * Manually flushing the {@link ChannelHandlerContext} is required for writes as the flow controller will
+     * not flush by itself.
```
Update the comment on line [85](https://github.com/netty/netty/pull/3712/files#diff-3ad0811d613c730f2815be14880147fdR85) to remove the `flush is required`.

Scottmitch replied:
nit: `not` -> `<strong>not</strong>`

normanmaurer replied:
+1

louiscryan replied:
Done

louiscryan replied:
Actually. I've just eliminated the return value completely as we never used it.

Scottmitch commented:
@louiscryan - LGTM.  What about the 2 remaining flushes for resetStream and goAway?  Can we do anything for those...seems like we don't want to miss a flush in those situations.

louiscryan commented:
I've removed the flushes from go-away/reset when called through the writer. I've added explicit flushes to cover go-away on close, go-away / reset streams on thrown exceptions.

louiscryan commented:
PTAL

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java:
```diff
@@ -434,6 +434,7 @@ public void onSettingsRead(ChannelHandlerContext ctx, Http2Settings settings) th
 
             // Acknowledge receipt of the settings.
             encoder.writeSettingsAck(ctx, ctx.newPromise());
+            ctx.flush();
```
Is this and the `onPingRead` just to be more responsive for setting up connections and responding to pings?

louiscryan replied:
Fixed, shouldn't have been there now that we flush on read complete.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java:
```diff
@@ -552,7 +558,9 @@ private ChannelFuture goAway(ChannelHandlerContext ctx, Http2Exception cause) {
         int lastKnownStream = connection().remote().lastStreamCreated();
-        return goAway(ctx, lastKnownStream, errorCode, debugData, ctx.newPromise());
+        ChannelFuture future = goAway(ctx, lastKnownStream, errorCode, debugData, ctx.newPromise());
+        ctx.flush();
```
The way the code is now we would should be OK to move this into `close(ctx, promise)` right?  However we have identified calling flush multiple times (which it seems like it will be if `onException` calls `onConnectionError` which calls `goAway`) so better safe than sorry for the future I guess? If this is your rational too maybe add a comment so we don't accidentally remove it?

louiscryan replied:
Moved into close, I think that's cleaner.

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java:
```diff
@@ -1109,26 +1107,21 @@ public Void answer(InvocationOnMock invocationOnMock) {
+        doAnswer(new Answer<Void>() {
+            @Override
+            public Void answer(InvocationOnMock invocationOnMock) throws Throwable {
+                size.addAndGet(-50);
```
Was this just to simplify, or required due to these changes?

Maybe it is better to have some variation?

louiscryan replied:
Variation won't really do anything as we no longer do any arithmetic based on the the number of bytes written by 'write' we just look to see if size == 0 after writing.

Scottmitch replied:
Ok no problem, but the tests don't necessarily have to change to match our implementation.  Just curious was the test failing without this change?

louiscryan replied:
Probably not, looking at it more closely. I probably could have just converted it to Answer<Void>

nmittler commented:
LGTM

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowControllerTest.java:
```diff
@@ -1176,25 +1169,25 @@ public Void answer(InvocationOnMock invocationOnMock) {
-                default:
-                    when(flowControlled.size()).thenReturn(10);
-                    throw new RuntimeException("Write failed");
+                    case 0:
```
@normanmaurer - Are we indenting or not?

normanmaurer replied:
the new style code says no....

louiscryan replied:
done

louiscryan commented:
Squashed commits and added appropriate commit message

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java:
```diff
@@ -552,7 +559,8 @@ private ChannelFuture goAway(ChannelHandlerContext ctx, Http2Exception cause) {
         long errorCode = cause != null ? cause.error().code() : NO_ERROR.code();
         ByteBuf debugData = Http2CodecUtil.toByteBuf(ctx, cause);
         int lastKnownStream = connection().remote().lastStreamCreated();
-        return goAway(ctx, lastKnownStream, errorCode, debugData, ctx.newPromise());
```
Revert this plz.

Also add comment to this method that a flush will <strong>not</strong> be done?

louiscryan replied:
Done.

Scottmitch commented:
@louiscryan - 1 more comment then will cherry-pick.

louiscryan commented:
All done

Scottmitch commented:
@louiscryan - Thanks for the updates!

Cherry-picked 4.1 (8271c8a) master (b122fae)
